exonum-java-binding/core/rust/src/storage/raw_proof_map_index.rs

```diff
@@ -0,0 +1,483 @@
+                Ok(env
+                    .new_object_unchecked(
+                        &iterWrapper.element_class,
+                        iterWrapper.constructor_id,
+                        &[key.into(), value.into()],
+                    )?
+                    .into_inner())
+            }
+            None => Ok(ptr::null_mut()),
+        }
+    });
+    utils::unwrap_exc_or(&env, res, ptr::null_mut())
+}
+
+/// Destroys the underlying `ProofMapIndex` iterator object and frees memory.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_core_storage_indices_RawProofMapIndexProxy_nativeEntriesIterFree(
+    env: JNIEnv,
+    _: JObject,
+    iter_handle: Handle,
+) {
+    handle::drop_handle::<Iter>(&env, iter_handle);
+}
+
+/// Returns the next value from the keys-iterator. Returns null pointer when iteration is finished.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_core_storage_indices_RawProofMapIndexProxy_nativeKeysIterNext(
+    env: JNIEnv,
+    _: JObject,
+    iter_handle: Handle,
+) -> jbyteArray {
+    let res = panic::catch_unwind(|| {
+        let iter = handle::cast_handle::<ProofMapIndexKeys<Key>>(iter_handle);
+        match iter.next() {
+            Some(val) => env.byte_array_from_slice(&val),
+            None => Ok(ptr::null_mut()),
+        }
+    });
+    utils::unwrap_exc_or(&env, res, ptr::null_mut())
+}
+
+/// Destroys the underlying `ProofMapIndex` keys-iterator object and frees memory.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_core_storage_indices_RawProofMapIndexProxy_nativeKeysIterFree(
+    env: JNIEnv,
+    _: JObject,
+    iter_handle: Handle,
+) {
+    handle::drop_handle::<ProofMapIndexKeys<Key>>(&env, iter_handle);
+}
+
+/// Return next value from the values-iterator. Returns null pointer when iteration is finished.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_core_storage_indices_RawProofMapIndexProxy_nativeValuesIterNext(
+    env: JNIEnv,
+    _: JObject,
+    iter_handle: Handle,
+) -> jbyteArray {
+    let res = panic::catch_unwind(|| {
+        let iter = handle::cast_handle::<ProofMapIndexValues<Value>>(iter_handle);
+        match iter.next() {
+            Some(val) => env.byte_array_from_slice(&val),
+            None => Ok(ptr::null_mut()),
+        }
+    });
+    utils::unwrap_exc_or(&env, res, ptr::null_mut())
+}
+
+/// Destroys the underlying `ProofMapIndex` values-iterator object and frees memory.
+#[no_mangle]
+pub extern "system" fn Java_com_exonum_binding_core_storage_indices_RawProofMapIndexProxy_nativeValuesIterFree(
+    env: JNIEnv,
+    _: JObject,
+    iter_handle: Handle,
+) {
+    handle::drop_handle::<ProofMapIndexValues<Value>>(&env, iter_handle);
+}
+
+// Converts Java byte array to key.
+fn convert_to_key(env: &JNIEnv, array: jbyteArray) -> JniResult<Key> {
+    let key = env.convert_byte_array(array)?;
+    assert_eq!(
+        key.len(),
+        PROOF_MAP_KEY_SIZE,
+        "Key size should be {} bits",
```

Comment:
```suggestion
        "Key size expected to be {} bits, found {} bits",
```

Reply:
Btw, why bits, not bytes?

Reply:
I'm totally OK with bytes but bits seem to be more traditional.

Reply:
That's an internal error, so it does not matter much (bits/bytes, it is checked in Java by

https://github.com/exonum/exonum-java-binding/blob/0b344d5f530666a3a8aa39faa7cb1b9a12aaab90/exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/indices/ProofMapKeyCheckingSerializerDecorator.java#L49-L53

, but both actual and expected shall be included.

Reply:
Ok, switched it to bytes - with both values printed this seems more suitable.